src/middleware.js

```diff
@@ -15,6 +14,160 @@ import { handleCarBlock } from './handlers/car-block.js'
  * @typedef {import('@web3-storage/gateway-lib').UnixfsContext} UnixfsContext
  */
 
+/**
+ *
+ * @param {string} s
+ * @returns {import('./bindings.js').TokenMetadata}
+ */
+function deserializeTokenMetadata (s) {
+  // TODO should this be dag-json?
+  return JSON.parse(s)
+}
+
+/**
+ *
+ * @param {import('./bindings.js').TokenMetadata} m
+ * @returns string
+ */
+function serializeTokenMetadata (m) {
+  // TODO should this be dag-json?
+  return JSON.stringify(m)
+}
+
+/**
+ *
+ * @param {Environment} env
+ * @param {import('@web3-storage/gateway-lib/handlers').CID} cid
+ * @returns {Promise<import('./constants.js').RATE_LIMIT_EXCEEDED>}
+ */
+async function checkRateLimitForCID (env, cid) {
+  const rateLimiter = env.RATE_LIMITER
+  if (!rateLimiter) {
+    console.warn('no rate limiter found')
+    return RATE_LIMIT_EXCEEDED.NO
+  }
+  const rateLimitResponse = await rateLimiter.limit({ key: cid.toString() })
+  if (rateLimitResponse.success) {
+    return RATE_LIMIT_EXCEEDED.NO
+  } else {
+    console.log(`limiting CID ${cid}`)
+    return RATE_LIMIT_EXCEEDED.YES
+  }
+}
+
+/**
+ *
+ * @param {Environment} env
+ * @param {string} authToken
+ * @returns TokenMetadata
+ */
+async function getTokenMetadata (env, authToken) {
+  const cachedValue = await env.AUTH_TOKEN_METADATA.get(authToken)
+  // TODO: we should implement an SWR pattern here - record an expiry in the metadata and if the expiry has passed, re-validate the cache after
+  // returning the value
+  if (cachedValue) {
+    return deserializeTokenMetadata(cachedValue)
+  } else {
+    const accounting = Accounting.create({ serviceURL: env.ACCOUNTING_SERVICE_URL })
```

Comment:
Maybe pass the cache and accounting API instace to this function? In general we should avoid passing everything everywhere since it makes it harder to test.

Reply:
The issue I see with passing the cache and accounting to this function is that we will always need to call the `Accounting.create...` to provide that instance in the function call, and if the value is in the cache already we don't even need to execute that call at all because it just decodes the cached value.

Reply:
Yes, understood but I'm not convinced it'll be a performance problem.

Related, I'd probably implement the cache as a transparent wrapper around the accounting service and pass the accounting service as a dependency to the rate limiting service.

i.e. something like:

```js
const accounting = Accounting.create({ serviceURL: env.ACCOUNTING_SERVICE_URL })
const cachedAccounting = Accounting.withCache(accounting, env.AUTH_TOKEN_METADATA)
const rateLimiter = RateLimiter.create(cachedAccounting)

const status = await rateLimiter.check(cid, request)
if (status  === RATE_LIMIT_EXCEEDED.YES) {
  // ...
```

So then it should be easier to test the rate limiter, since the rate limiter is not importing a function that requires a KV _and_ an accounting service. Also easier to test the caching logic in isolation.

...anyways...just a suggestion - non blocking.